### PR TITLE
Nano: support evaluation related features for INC

### PR DESCRIPTION
**Pull request by rnwang04**
## Description

After we have upgrade INC version to 2.0, we have not supported `eval_metric` for this version, and we never support `eval function` for quantization until now, which is an effective tool to support a more flexible evaluation form. However, `eval_metric` and `eval_function` are important for the quantization quality.

### 1. Why the change?

To support `eval_metric` and `eval_function` in `InferenceOptimizer.quantize` for both PyTorch and Tensorflow.

### 2. User API changes

new parameter `eval_function` for `InferenceOptimizer.quantize`, here take PyTorch for example:
```python
def eval_func(model):
    acc = 0
    for sample, label in self.train_loader:
        logits = model(sample)
        pred = logits.argmax(dim=1)
        acc += torch.eq(pred, label).sum().float().item()
    return acc

qmodel = InferenceOptimizer.quantize(pl_model,
                                     calib_data=self.train_loader,
                                     eval_func=eval_func,
                                     accuracy_criterion={'relative': 0.8,
                                                         'higher_is_better': True},
                                     timeout=0,
                                     max_trials=10,
                                     thread_num=8)
```
At the same time, `metric` parameter now actually works for INC 2.0 version.
### 3. Summary of the change 

-  support eval function for INC 2.0 in PyTorch InferenceOptimizer
- support eval function for INC 2.0 in keras InferenceOptimizer
- support eval function for INC 1.x in PyTorch InferenceOptimizer
- support eval function for **INC 1.14.1 or higher version** in keras InferenceOptimizer
- support eval metric for INC 2.0 in PyTorch InferenceOptimizer
- support eval metric for INC 2.0 in keras InferenceOptimizer
- related uts for PyTorch
- related uts for keras
- update comment for `use_ipex` in PyTorch InferenceOptimizer
- update comment for `accuracy_criterion` in `quantize`

### 4. How to test?
- [x] Unit test

### CODE REVIEW

File: python/nano/src/bigdl/nano/deps/neural_compressor/inc_api_2.py [hunk 1]
```diff
@@ -93,6 +97,7 @@ def _quantize(
     from neural_compressor.conf.pythonic_config import Config
     from neural_compressor.conf.config import QuantConf
     from neural_compressor.data import DataLoader
+    from neural_compressor.experimental.common import Metric as NCMetric
```
Comment (MeouSker77) [inline python/nano/src/bigdl/nano/deps/neural_compressor/inc_api_2.py:100]: why not `from neural_compressor.metric import Metric`

File: python/nano/src/bigdl/nano/deps/neural_compressor/inc_api_2.py [hunk 2]
```diff
@@ -144,8 +191,7 @@ def new_collate_fn(batch):
         accuracy_criterion = AccuracyCriterion(
             higher_is_better=accuracy_criterion.get('higher_is_better', True),
             criterion=criterion,
-            # we calculate `tolerable_loss` by `1 - target_accuracy`
-            tolerable_loss=1.0 - accuracy_criterion[criterion]
+            tolerable_loss=float(accuracy_criterion[criterion])
```
Comment (MeouSker77) [inline python/nano/src/bigdl/nano/deps/neural_compressor/inc_api_2.py:194]: shall we rename `accuracy_criterion` to `loss_criterion` ?

File: python/nano/src/bigdl/nano/deps/neural_compressor/inc_api_2.py (hunk 1 above)
Comment (rnwang04) [reply python/nano/src/bigdl/nano/deps/neural_compressor/inc_api_2.py:100]: This import is useless now, will delete this line.

File: python/nano/src/bigdl/nano/deps/neural_compressor/inc_api_2.py (hunk 2 above)
Comment (rnwang04) [reply python/nano/src/bigdl/nano/deps/neural_compressor/inc_api_2.py:194]: I think we can still use `accuracy_criterion` as we have detailed comment `:param accuracy_criterion:  Tolerable accuracy drop, defaults to None meaning no accuracy control.`

File: python/nano/src/bigdl/nano/deps/neural_compressor/core/quantization.py [hunk 3]
```diff
@@ -49,6 +50,11 @@ def __init__(self,
                                      allows relative accuracy loss: 1%. accuracy_criterion = {
```
Comment (MeouSker77) [inline python/nano/src/bigdl/nano/deps/neural_compressor/core/quantization.py:50]: may update this comment, for my understanding, `'relative': 0.1` means accuracy >= 10% * original_accuracy ? and `'absolute: 0.99'` means accuracy > 0.99 ?

Comment (rnwang04) [reply python/nano/src/bigdl/nano/deps/neural_compressor/core/quantization.py:50]: For exmaple, if `'absolute: 0.99'` and `higher_is_better` is True, this requires original metric value subtract current metric value be smaller than 0.99. For `'relative'`, it's similar.
Have revised corresponding comment.

Comment (MeouSker77) [reply python/nano/src/bigdl/nano/deps/neural_compressor/core/quantization.py:50]: Oh, I understand. But why we set it to `relative: 0.99` by default, this allows 99% metric value loss ? INC's default value is `relative: 0.01`, why we use 0.99 ?

Comment (rnwang04) [reply python/nano/src/bigdl/nano/deps/neural_compressor/core/quantization.py:50]: I think `relative: 0.99` actually means no tunning, and we set this value to ensure quantization always works and don't raise error like `can't find model meeting accuracy_criterion`.
I have tried` relative: 0.01` before and some ut of chronos will failed as can't get quantized model meet such strict accuracy limit.

Comment (MeouSker77) [reply python/nano/src/bigdl/nano/deps/neural_compressor/core/quantization.py:50]: well, but I think 0.99 is too relaxed, if the metric value drops to 1%, I think this optimization is pointless. Have we tried other value? such as 0.9, 0.8 ?

Comment (rnwang04) [reply python/nano/src/bigdl/nano/deps/neural_compressor/core/quantization.py:50]: `if the metric value drops to 1%, I think this optimization is pointless.`
Of course, in most cases, this will not happen.
However, in our uts, we always use random data for quantization, so that the accuracy of quantized model has no guarantee.
I will try 0.8 but such value will import a potential risk for action failing.

Comment (MeouSker77) [reply python/nano/src/bigdl/nano/deps/neural_compressor/core/quantization.py:50]: > I will try 0.8 but such value will import a potential risk for action failing.

We can set its default value to 0.01, as for the UT, we can pass 0.99 to this parameter

Comment (rnwang04) [reply python/nano/src/bigdl/nano/deps/neural_compressor/core/quantization.py:50]: Or there is another solution, we set default value to 0.01 as INC's, and adjust related failed uts. How do you think of?

Comment (MeouSker77) [reply python/nano/src/bigdl/nano/deps/neural_compressor/core/quantization.py:50]: well, we have the same idea

**Review by MeouSker77 — APPROVED** [review]
LGTM
